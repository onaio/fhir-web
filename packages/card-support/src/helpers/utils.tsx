/** Function to download data to a file
 *
 * @param {string} data - data to be written to file
 * @param {string} filename - name of the file to be saved
 * @param {string} type - MIME type for the file
 */
export const downloadFile = (data: string, filename: string, type: string) => {
  const file = new Blob([data], { type });
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (window.navigator.msSaveOrOpenBlob) {
    // IE10+
    window.navigator.msSaveOrOpenBlob(file, filename);
  } else {
    // Others
    const docElement = document.createElement('a');
    const url = URL.createObjectURL(file);
    docElement.href = url;
    docElement.download = filename;
    document.body.appendChild(docElement);
    docElement.click();
    setTimeout(() => {
      document.body.removeChild(docElement);
      window.URL.revokeObjectURL(url);
    }, 0);
  }
};
