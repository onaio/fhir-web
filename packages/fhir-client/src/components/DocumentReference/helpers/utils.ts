import { get } from 'lodash';

/**
 * get object from an array, the check parameters are passed in as an obj, keys in keyValue can represent a path
 *
 * @param objArray - array obj with records of objs
 * @param keyValue - obj defining what key and values to use when matching
 * @param all - return first matched or all
 */
export function findObj<T extends object>(
  objArray?: T[],
  keyValue?: Record<keyof T | string, unknown>,
  all = false
) {
  if (objArray && keyValue) {
    const checkParameters = Object.entries(keyValue);
    const matched = objArray.filter((obj) => {
      let matches = true;
      checkParameters.forEach(([key, value]) => {
        if (get(obj, key) !== value) {
          matches = false;
        }
      });
      return matches;
    });
    if (all) {
      return matched;
    } else {
      return matched[0];
    }
  }
  if (!objArray || !keyValue) return;
}
