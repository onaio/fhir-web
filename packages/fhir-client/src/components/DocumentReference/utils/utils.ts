import { get, isEmpty, some } from 'lodash';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { format } from 'date-fns';
import { dateFormat } from '../../../constants';
import FHIR from 'fhirclient';

/**
 * checks if collection is empty or if all values in the obj can be said to hold null values
 *
 * @param coll - the collection
 */
export const itemIsEmpty = <T extends object>(coll?: T) => {
  if (isEmpty(coll)) {
    return true;
  }
  return !some(coll);
};

/**
 * process top level fields of the documentReference resource
 *
 * @param docResource DocumentResource
 */
export const processTopLevelFields = (docResource: IfhirR4.IDocumentReference) => {
  const { status, type, description, securityLabel, date, id } = docResource;
  return {
    id,
    title: description,
    createdAt: date ? format(new Date(date), dateFormat) : undefined,
    status,
    documentType: {
      codeList: type?.coding,
      code: get(type, 'coding.0'),
    },
    securityCodes: {
      codeList: securityLabel,
      code: get(securityLabel, '[0].coding[0]'),
    },
  };
};

/**
 * Process context field in a documentReference
 *
 * @param docResource - the document resource
 */
export const processContextFields = (docResource: IfhirR4.IDocumentReference) => {
  const start = get(docResource, 'context.period.start');
  const end = get(docResource, 'context.period.end');
  return {
    periodStart: start ? format(new Date(start), dateFormat) : undefined,
    periodEnd: end ? format(new Date(end), dateFormat) : undefined,
    eventCoding: get(docResource, 'context.event[0].coding[0]'),
    facilityTypeCoding: get(docResource, 'context.facilityType.coding[0]'),
    practiceSettingCoding: get(docResource, 'context.practiceSetting.coding[0]'),
  };
};

/**
 * extract values from content field, will include the attachments and their formats
 *
 * @param docResource - the document resource
 */
export const processContentFields = (docResource: IfhirR4.IDocumentReference) => {
  return get(docResource, 'content').map((content) => {
    const data = get(content, 'attachment.data');
    const url = get(content, 'attachment.url');
    const absoluteUrlRegex = /^(?:[a-z]+:)?\/\//;
    const binaryUrlRegex = /Binary\/[\w-]+$/gm;
    let urlIsAbsolute, binaryUrl;
    if (url) {
      urlIsAbsolute = absoluteUrlRegex.test(url);
      const matches = Array.from(
        url.matchAll(binaryUrlRegex),
        (m) => (m as Array<string | undefined>)[0]
      );
      binaryUrl = matches[0];
    }
    const size = get(content, 'attachment.size');
    const formatCode = get(content, 'format.code');
    const formatDisplay = get(content, 'format.display');
    const formatContentType = get(content, 'attachment.contentType');

    return {
      attachment: { url, urlIsAbsolute, binaryUrl, data },
      size,
      formatCode,
      formatDisplay,
      formatContentType,
    };
  });
};

/**
 * parses docResources and extracts attachment data, filtering out those with empty string data uris
 *
 * @param docResources - documentReference resource array
 */
export const processDocumentReferences = (docResources: IfhirR4.IDocumentReference[]) => {
  return docResources.map((resource) => {
    return {
      ...processTopLevelFields(resource),
      context: processContextFields(resource),
      content: processContentFields(resource),
    };
  });
};

export type ParsedDocReference = ReturnType<typeof processDocumentReferences>;

/**
 * get the discrete and multipart of a content-type
 *
 * @param contentType - mime type
 */
export const splitContentType = (contentType: string) => {
  const splitText = contentType.split('/');
  return { discretePart: splitText[0], multipart: splitText[1] };
};

/**
 * Act as a proxy when downloading attachments so that we can customize the name.
 * This pulls attachment as blob from fhir /Binary endpoint
 *
 * @param fhirBaseURL -the fhir base url
 * @param binaryEndpoint - a Binary endpoint url path
 */
export const fetchAttachmentForDownload = async (fhirBaseURL: string, binaryEndpoint: string) => {
  return FHIR.client(fhirBaseURL)
    .request({
      url: binaryEndpoint,
    })
    .then((res) => {
      return res.blob();
    });
};
