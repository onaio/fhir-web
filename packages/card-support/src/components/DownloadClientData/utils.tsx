import Papaparse from 'papaparse';
import { OpenSRPService } from '@opensrp/server-service';
import { DownloadClientDataFormFields } from '../DownloadClientData';
import { Dispatch, SetStateAction } from 'react';
import { OPENSRP_URL_CLIENT_SEARCH, APPLICATION_CSV } from '../../constants';
import { Client } from '../../ducks/clients';
import { downloadFile } from '../../helpers/utils';
import { ParsedHierarchyNode } from '@opensrp/location-management';
import { sendErrorNotification } from '@opensrp/notifications';
import { Dictionary } from '@onaio/utils';
import lang, { Lang } from '../../lang';
/* eslint-disable @typescript-eslint/camelcase */

/** interface for user assignment response */
export interface UserAssignment {
  organizationIds: number[];
  jurisdictions: string[];
  plans: string[];
}

/** interface for CSV entries */
export interface ClientCSVEntry {
  id: string;
  dob: string;
  first_name: string;
  last_name: string;
  gender: string;
  facility_of_registration: string;
  date_of_registration: string;
  card_status: string;
  card_status_last_update: string;
}

/**
 * Handle form date change
 *
 * @param {string} range - selected date range
 * @param {Dispatch<SetStateAction<[string, string]>>} setCardOrderDate - state dispatcher
 */
export const handleCardOrderDateChange = (
  range: [string, string],
  setCardOrderDate: Dispatch<SetStateAction<[string, string]>>
): void => {
  setCardOrderDate(range);
};

/**
 * Handle download client data form submission
 *
 * @param values - sumitted form values
 * @param accessToken - OPENSRP API access token
 * @param opensrpBaseURL - OPENSRP API base URL
 * @param serviceClass - OPENSRP service class
 * @param locations location hierarchy
 * @param setSubmitting - method to set form `isSubmitting` status
 * @param langObj - tthe lang object
 */
export const submitForm = (
  values: DownloadClientDataFormFields,
  accessToken: string,
  opensrpBaseURL: string,
  serviceClass: typeof OpenSRPService,
  locations: ParsedHierarchyNode[],
  setSubmitting: (isSubmitting: boolean) => void,
  langObj: Lang = lang
): void => {
  const { clientLocation, cardStatus, cardOrderDate } = values;

  if (!clientLocation) {
    return;
  }

  setSubmitting(true);
  const startDate = cardOrderDate[0];
  const endDate = cardOrderDate[1];
  const endPoint = `${OPENSRP_URL_CLIENT_SEARCH}`;
  const serve = new serviceClass(accessToken, opensrpBaseURL, endPoint);
  let params: Dictionary = {
    locationIds: clientLocation,
  };

  if (cardStatus) {
    params = {
      ...params,
      attribute: `card_status:${cardStatus}`,
    };
  }

  serve
    .list(params)
    .then((clients: Client[]) => {
      const location = getLocationDetails(locations, clientLocation);
      const entries: ClientCSVEntry[] = clients
        .filter((client: Client) => !!client.identifiers.zeir_id)
        .filter((client: Client) => {
          const dateCreated = new Date(client.dateCreated.split('T')[0]);

          return (
            dateCreated.getTime() >= new Date(startDate).getTime() &&
            dateCreated.getTime() <= new Date(endDate).getTime()
          );
        })
        .map((client: Client) => {
          return {
            id: client.identifiers.zeir_id,
            dob: formatDDMMYYY(new Date(client.birthdate.split('T')[0])),
            first_name: client.firstName,
            last_name: client.lastName,
            gender: client.gender,
            facility_of_registration: location ? location.title : '',
            date_of_registration: formatDDMMYYY(new Date(client.dateCreated.split('T')[0])),
            card_status: client.attributes.card_status,
            card_status_last_update: formatDDMMYYY(
              new Date(client.attributes.card_status_date.split('T')[0])
            ),
          };
        });

      setSubmitting(false);

      if (entries.length) {
        createCsv(entries, buildCSVFileName(location ? location.title : '', startDate, endDate));
      } else {
        sendErrorNotification(langObj.NO_DATA_FOUND);
      }
    })
    .catch((_: Error) => {
      setSubmitting(false);
      sendErrorNotification(langObj.ERROR_OCCURRED);
    });
};

/**
 * Format a Date object as dd MM YYYY
 *
 * @param {Date} date - date object
 * @param {string} delimiter - delimiter to use to separate the different parts
 * @returns {string} formatted date
 */
export const formatDDMMYYY = (date: Date, delimiter = '/'): string => {
  //eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  return `${('0' + date.getDate()).slice(-2)}${delimiter}${('0' + (date.getMonth() + 1)).slice(
    -2
  )}${delimiter}${date.getFullYear()}`;
};

/**
 * Build the CSV file name
 *
 * @param {string} location - name of the location filter
 * @param {string} startDate - start date of the range filter
 * @param {string} endDate - end date of the range filter
 * @returns {string} CSV file name
 */
export const buildCSVFileName = (location: string, startDate: string, endDate: string): string => {
  return `Children_list_${location}_${formatDDMMYYY(new Date(), '_')}_(${formatDDMMYYY(
    new Date(startDate),
    '-'
  )} - ${formatDDMMYYY(new Date(endDate), '-')})`;
};

// Create csv data for the selected jurisdiction hieracrhy
export const createCsv = (entries: ClientCSVEntry[], fileName: string): void => {
  const csv: string = Papaparse.unparse(entries, {
    header: true,
  });
  // Export csv file
  downloadFile(csv, `${fileName}.csv`, APPLICATION_CSV);
};

/**
 * Get location from the location hierarchy
 *
 * @param {ParsedHierarchyNode} locations location hierarchy
 * @param {string} locationId id of location to return
 * @returns {ParsedHierarchyNode | null} location if found, null if not found
 */
export const getLocationDetails = (
  locations: ParsedHierarchyNode[],
  locationId: string
): ParsedHierarchyNode | null => {
  let found = false;
  let i = 0;
  let location = null;

  while (!found && i < locations.length) {
    const node = locations[i];

    if (node.id && node.id === locationId) {
      found = true;
      location = node;
    } else if (node.children) {
      location = getLocationDetails(node.children, locationId);

      if (location) {
        found = true;
      }
    }

    i += 1;
  }

  return location;
};
