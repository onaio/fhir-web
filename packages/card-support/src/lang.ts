import i18n, { namespace } from './mls';
import { Dictionary } from '@onaio/utils';

export type Lang = Dictionary<string>;
const lang: Lang = {};

/** recompute values */
function fill() {
  lang.ERROR_OCCURRED = i18n.t(`An error occurred`, { ns: namespace });
  lang.DOWNLOAD_CLIENT_DATA = i18n.t(`Download Client Data`, { ns: namespace });
  lang.CLIENT_LOCATION = i18n.t(`Client Location`, { ns: namespace });
  lang.ALL_LOCATIONS = i18n.t(`All Locations`, { ns: namespace });
  lang.CARD_STATUS = i18n.t(`Card Status`, { ns: namespace });
  lang.NEEDS_CARD = i18n.t(`Needs Card`, { ns: namespace });
  lang.CARD_NOT_NEEDED = i18n.t(`Card not needed`, { ns: namespace });
  lang.BOTH_NEEDS_CARD_CARD_NOT_NEEDED = i18n.t(`Both "Needs card" and "Card not needed"`, {
    ns: namespace,
  });
  lang.CARD_ORDER_DATE = i18n.t(`Card Order Date`, { ns: namespace });
  lang.CARD_ORDER_DATE_REQUIRED = i18n.t(`Please enter start date and end date`, { ns: namespace });
  lang.SELECT_CARD_ORDER_DATE = i18n.t(`Select Card Order Date`, { ns: namespace });
  lang.DOWNLOADING = i18n.t(`Downloading`, { ns: namespace });
  lang.DOWNLOAD_CSV = i18n.t(`Download CSV`, { ns: namespace });
  lang.NO_DATA_FOUND = i18n.t(`No data found`, { ns: namespace });
}

// run it initial
fill();

// bind some events and fill values again (doing the magic you expect to happen magically)
i18n.on(`languageChanged`, () => {
  fill();
});

// export the const
export default lang;
