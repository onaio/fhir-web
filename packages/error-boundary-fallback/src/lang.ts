import i18n, { namespace } from './mls';
import { Dictionary } from '@onaio/utils';

const lang: Dictionary<string> = {};

/** recompute values */
function fill() {
  lang.PAGE_TITLE = i18n.t(`An Error Occurred`, { ns: namespace });
  lang.PAGE_SUB_TITLE = i18n.t(
    `There has been an error. Itâ€™s been reported to the site administrators via email and should be fixed shortly. Thanks for your patience.'`,
    { ns: namespace }
  );
  lang.BUTTON_TITLE = i18n.t(`Back Home`, { ns: namespace });
}

// run it initial
fill();

// bind some events and fill values again (doing the magic you expect to happen magically)
i18n.on(`languageChanged`, () => {
  fill();
});

// export the const
export default lang;
