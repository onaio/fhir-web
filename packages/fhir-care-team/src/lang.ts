import i18n, { namespace } from './mls';
import { Dictionary } from '@onaio/utils';

export type Lang = Dictionary<string>;

const lang: Lang = {};

/** recompute values */
function fill() {
  // Tooltips
  lang.TIP_REQUIRED_FIELD = i18n.t(`This is a required field`, { ns: namespace });

  // Errors
  lang.ERROR_OCCURRED = i18n.t(`An error occurred`, { ns: namespace });

  // Messages
  lang.CARE_TEAMS_UPDATE_SUCCESS = i18n.t(`Successfully Updated Care Teams`, { ns: namespace });
  lang.CARE_TEAMS_ADD_SUCCESS = i18n.t(`Successfully Added Care Teams`, { ns: namespace });
  lang.CARE_TEAM_DELETE_SUCCESS = i18n.t(`Successfully Deleted Care Team`, { ns: namespace });
  lang.CONFIRM_DELETE = i18n.t(`Are you sure you want to delete this Care Team?`, {
    ns: namespace,
  });
  lang.YES = i18n.t(`Yes`, { ns: namespace });
  lang.NO = i18n.t(`No`, { ns: namespace });

  // Rendered text
  lang.STATUS = i18n.t(`Status`, { ns: namespace });
  lang.NAME = i18n.t(`Name`, { ns: namespace });
  lang.SAVING = i18n.t(`Saving`, { ns: namespace });
  lang.SAVE = i18n.t(`Save`, { ns: namespace });
  lang.ACTIVE = i18n.t(`Active`, { ns: namespace });
  lang.INACTIVE = i18n.t(`Inactive`, { ns: namespace });
  lang.CANCEL = i18n.t(`Cancel`, { ns: namespace });
  lang.EDIT_CARE_TEAM = i18n.t(`Edit Care Team`, { ns: namespace });
  lang.CREATE_CARE_TEAM = i18n.t(`Create Care Team`, { ns: namespace });
  lang.EDIT = i18n.t(`Edit`, { ns: namespace });
  lang.CREATE = i18n.t(`Create`, { ns: namespace });
  lang.IDENTIFIER = i18n.t(`Identifier`, { ns: namespace });
  lang.TEAMS = i18n.t(`Care Teams`, { ns: namespace });
  lang.CARE_TEAM_MEMBERS = i18n.t(`Care Teams`, { ns: namespace });
  lang.SEARCH = i18n.t(`Search`, { ns: namespace });
  lang.IDENTIFIER = i18n.t(`Identifier`, { ns: namespace });
  lang.CARE_TEAM_PAGE_HEADER = i18n.t(`FHIR Care Team`, { ns: namespace });
  lang.SUBJECT = i18n.t(`Subject`, { ns: namespace });
  lang.PARTICIPANTS = i18n.t(`Participant`, { ns: namespace });
}

// run it initial
fill();

// bind some events and fill values again (doing the magic you expect to happen magically)
i18n.on(`languageChanged`, () => {
  fill();
});

// export the const
export default lang;
