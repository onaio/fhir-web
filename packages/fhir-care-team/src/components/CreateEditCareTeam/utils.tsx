import { history } from '@onaio/connected-reducer-registry';
import { v4 } from 'uuid';
import FHIR from 'fhirclient';
import { sendErrorNotification, sendSuccessNotification } from '@opensrp/notifications';
import lang from '../../lang';
import { URL_CARE_TEAM } from '../../constants';
import { Dictionary } from '@onaio/utils';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { FormFields } from './Form';

export const submitForm = async (
  values: FormFields,
  fhirBaseURL: string,
  id?: string,
  uuid?: string
): Promise<void> => {
  const careTeamId = uuid ? uuid : v4();
  const payload: Omit<IfhirR4.ICareTeam, 'meta'> = {
    resourceType: 'CareTeam',
    identifier: [
      {
        use: 'official',
        value: careTeamId,
      },
    ],
    id: id,
    name: values.name,
    status: values.status as IfhirR4.CareTeam.StatusEnum,
    subject: {
      reference: `Group/${values.groupsId}`,
    },
    participant: values.practitionersId?.map((id) => {
      return {
        member: {
          reference: `Practitioner/${id}`,
        },
      };
    }),
  };
  const serve = FHIR.client(fhirBaseURL);
  if (id) {
    await serve
      .update(payload)
      .then(() => sendSuccessNotification(lang.CARE_TEAMS_UPDATE_SUCCESS))
      .catch(() => sendErrorNotification(lang.ERROR_OCCURRED));
  } else {
    await serve
      .create(payload)
      .then(() => sendSuccessNotification(lang.CARE_TEAMS_ADD_SUCCESS))
      .catch(() => sendErrorNotification(lang.ERROR_OCCURRED));
  }
  history.push(URL_CARE_TEAM);
};

/** Util function to build out patient or practitioner name
 *
 * @param {object} patient - patient resource object
 * @returns {string} - returns patient or practitioner name string
 */
export function getPatientName(patient: Dictionary | undefined) {
  if (!patient) {
    return '';
  }

  let name = patient.name;
  if (!Array.isArray(name)) {
    name = [name];
  }
  name = name[0];
  if (!name) {
    return '';
  }

  const family = Array.isArray(name.family) ? name.family : [name.family];
  const given = Array.isArray(name.given) ? name.given : [name.given];
  const prefix = Array.isArray(name.prefix) ? name.prefix : [name.prefix];
  const suffix = Array.isArray(name.suffix) ? name.suffix : [name.suffix];

  return [
    prefix.map((t = '') => String(t).trim()).join(' '),
    given.map((t = '') => String(t).trim()).join(' '),
    family.map((t = '') => String(t).trim()).join(' '),
    suffix.map((t = '') => String(t).trim()).join(' '),
  ]
    .filter(Boolean)
    .join(' ');
}
