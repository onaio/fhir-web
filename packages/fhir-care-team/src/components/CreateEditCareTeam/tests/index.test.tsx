import { mount, shallow } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { history } from '@onaio/connected-reducer-registry';
import { Router } from 'react-router';
import * as reactQuery from 'react-query';
import * as fixtures from './fixtures';
import * as fhirCient from 'fhirclient';
import flushPromises from 'flush-promises';
import { createTestQueryClient } from '../../ListView/tests/utils';
import * as notifications from '@opensrp/notifications';
import { defaultInitialValues, CreateEditCareTeam } from '..';
import toJson from 'enzyme-to-json';
import lang from '../../../lang';

const { QueryClientProvider } = reactQuery;

const testQueryClient = createTestQueryClient();

/* eslint-disable @typescript-eslint/camelcase */

jest.mock('@opensrp/notifications', () => ({
  __esModule: true,
  ...Object.assign({}, jest.requireActual('@opensrp/notifications')),
}));

describe('components/CreateEditCareTeam', () => {
  const props = {
    history,
    fhirBaseURL: 'https://r4.smarthealthit.org/',
    location: {
      hash: '',
      pathname: '/CareTeam/edit',
      search: '',
      state: '',
    },
    match: {
      isExact: true,
      params: { careTeamId: fixtures.careTeam1.id },
      path: `/CareTeam/edit/:careTeamId`,
      url: `/CareTeam/edit/${fixtures.careTeam1.id}`,
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    act(() => {
      shallow(
        <Router history={history}>
          <QueryClientProvider client={testQueryClient}>
            <CreateEditCareTeam {...props} />
          </QueryClientProvider>
        </Router>
      );
    });
  });

  it('renders correctly', async () => {
    const fhir = jest.spyOn(fhirCient, 'client');
    fhir.mockImplementation(
      jest.fn().mockImplementation(() => {
        return {
          request: jest.fn().mockResolvedValueOnce(fixtures.careTeam1),
        };
      })
    );

    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <CreateEditCareTeam {...props} />
        </QueryClientProvider>
      </Router>
    );

    expect(toJson(wrapper.find('.ant-spin'))).toBeTruthy();

    await act(async () => {
      await flushPromises();
    });

    wrapper.update();

    const row = wrapper.find('Row').at(0);

    expect(row.text()).toMatchSnapshot('full care tem form');

    wrapper.unmount();
  });

  it('renders correctly for create care team', async () => {
    const propsCreate = {
      history,
      fhirBaseURL: 'https://r4.smarthealthit.org/',
      location: {
        hash: '',
        pathname: '/CareTeam/new',
        search: '',
        state: '',
      },
      match: {
        isExact: true,
        params: { careTeamId: '' },
        path: `/CareTeam/new`,
        url: `/CareTeam/new`,
      },
    };

    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <CreateEditCareTeam {...propsCreate} />
        </QueryClientProvider>
      </Router>
    );

    await act(async () => {
      await flushPromises();
    });

    wrapper.update();

    const row = wrapper.find('Row').at(0);

    expect(row.find('CareTeamForm').prop('initialValues')).toEqual(defaultInitialValues);

    wrapper.unmount();
  });

  it('fetches care team if page is refreshed', async () => {
    const fhir = jest.spyOn(fhirCient, 'client');
    fhir.mockImplementation(
      jest.fn().mockImplementation(() => {
        return {
          request: jest.fn().mockResolvedValue(fixtures.careTeam1),
        };
      })
    );
    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <CreateEditCareTeam {...props} />
        </QueryClientProvider>
      </Router>
    );

    await act(async () => {
      await flushPromises();
    });

    wrapper.update();

    // check if form initial values are set

    expect(wrapper.find('CareTeamForm').props().initialValues).toEqual({
      groupsId: '306',
      id: '308',
      name: 'Care Team One',
      practitionersId: ['206', '103'],
      status: 'active',
      uuid: '93bc9c3d-6321-41b0-9b93-1275d7114e22',
    });

    wrapper.unmount();
  });

  it('handles error if fetch fails when page reloads', async () => {
    const mockNotificationError = jest.spyOn(notifications, 'sendErrorNotification');

    const fhir = jest.spyOn(fhirCient, 'client');
    fhir.mockImplementation(
      jest.fn().mockImplementation(() => {
        return {
          request: jest.fn().mockRejectedValue('Failed'),
        };
      })
    );

    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <CreateEditCareTeam {...props} />
        </QueryClientProvider>
      </Router>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(mockNotificationError).toHaveBeenCalledWith(lang.ERROR_OCCURED);
    wrapper.unmount();
  });
});
