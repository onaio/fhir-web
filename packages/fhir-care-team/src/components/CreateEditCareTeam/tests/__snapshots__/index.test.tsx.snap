// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CreateEditCareTeam renders correctly: full care tem form 1`] = `"Edit Care Team | Care Team OneUUIDNameStatusActiveInactiveParticipant206103Â Subject306SaveCancel"`;
